stan/math/prim/prob/beta_lpdf.hpp

```diff
@@ -60,122 +65,79 @@ return_type_t<T_y, T_scale_succ, T_scale_fail> beta_lpdf(
   }
 
   T_partials_return logp(0);
-  operands_and_partials<T_y, T_scale_succ, T_scale_fail> ops_partials(y, alpha,
-                                                                      beta);
-  scalar_seq_view<T_y> y_vec(y);
-  scalar_seq_view<T_scale_succ> alpha_vec(alpha);
-  scalar_seq_view<T_scale_fail> beta_vec(beta);
-  size_t size_y = stan::math::size(y);
-  size_t size_alpha = stan::math::size(alpha);
-  size_t size_beta = stan::math::size(beta);
-  size_t N = max_size(y, alpha, beta);
+  operands_and_partials<T_y_ref, T_alpha_ref, T_beta_ref> ops_partials(
+      y_ref, alpha_ref, beta_ref);
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return y_dbl = value_of(y_vec[n]);
-    if (y_dbl < 0 || y_dbl > 1) {
-      return LOG_ZERO;
-    }
-  }
+  const auto& y_col = as_column_vector_or_scalar(y_ref);
+  const auto& alpha_col = as_column_vector_or_scalar(alpha_ref);
+  const auto& beta_col = as_column_vector_or_scalar(beta_ref);
 
-  VectorBuilder<include_summand<propto, T_y, T_scale_succ>::value,
-                T_partials_return, T_y>
-      log_y(size_y);
-  VectorBuilder<include_summand<propto, T_y, T_scale_fail>::value,
-                T_partials_return, T_y>
-      log1m_y(size_y);
+  const auto& y_arr = as_array_or_scalar(y_col);
+  const auto& alpha_arr = as_array_or_scalar(alpha_col);
+  const auto& beta_arr = as_array_or_scalar(beta_col);
 
-  for (size_t n = 0; n < size_y; n++) {
-    if (include_summand<propto, T_y, T_scale_succ>::value) {
-      log_y[n] = log(value_of(y_vec[n]));
-    }
-    if (include_summand<propto, T_y, T_scale_fail>::value) {
-      log1m_y[n] = log1m(value_of(y_vec[n]));
-    }
-  }
+  ref_type_t<decltype(value_of(y_arr))> y_val = value_of(y_arr);
+  ref_type_t<decltype(value_of(alpha_arr))> alpha_val = value_of(alpha_arr);
+  ref_type_t<decltype(value_of(beta_arr))> beta_val = value_of(beta_arr);
```

Comment:
Why go through:
  - ```Ref```
  - ```as_column_or_scalar```
  - ```as_array_or_scalar```
  - ```value_of```

Why not just ```as_array_or_scalar(value_of(x))```?

Reply:
ref is used in operands_and_partials so we can not skip it. If we skip `as_column_vector_or_scalar` we could end up with a row vector and its dimensions won't match other arguments. 

Now that holder is merged into develop I could use whole sequence of calls in one line. This would make some dynamic allocations this approach avoids, but would be a bit less verbose. I am not sure which approach is better here.